### PR TITLE
Add LinearGradientBrush tests and fix compat with GDI+

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

File: src/lineargradientbrush.c [hunk 1]
```diff
@@ -444,7 +444,7 @@ GdipCreateLineBrushI (GDIPCONST GpPoint *point1, GDIPCONST GpPoint *point2, ARGB
 {
 	GpPointF p1, p2;
 
-	if (!point1 || !point2 || !lineGradient)
+	if (!point1 || !point2 || !lineGradient || wrapMode == WrapModeClamp)
```
Comment (hughbe) [inline src/lineargradientbrush.c]: GDI+ rejects WrapModeClamp

File: src/lineargradientbrush.c [hunk 2]
```diff
@@ -555,23 +555,30 @@ GdipCreateLineBrushFromRectI (GDIPCONST GpRect *rect, ARGB color1, ARGB color2,
...
+
 	rectf.X = rect->X;
 	rectf.Y = rect->Y;
 	rectf.Width = rect->Width;
 	rectf.Height = rect->Height;
 
-	/* FIXME: Check whether angle has to be scalable or not in case of lineargradient mode. */
 	return GdipCreateLineBrushFromRectWithAngle (&rectf, color1, color2,
-		get_angle_from_linear_gradient_mode (mode), FALSE, wrapMode, lineGradient);
+		get_angle_from_linear_gradient_mode (mode), TRUE, wrapMode, lineGradient);
```
Comment (hughbe) [inline src/lineargradientbrush.c]: The angle has to be scalable - the tests pass against GDI+ but fail with libgdiplus without this change

File: src/lineargradientbrush.c [hunk 3]
```diff
@@ -628,13 +638,6 @@ GdipGetLineBlendCount (GpLineGradient *brush, int *count)
 	if (!brush || !count)
 		return InvalidParameter;
 
-	/* If count is less than 1, we are not in a proper state
```
Comment (hughbe) [inline src/lineargradientbrush.c]: GDI+ allows getting the count if there are no blends

File: src/lineargradientbrush.c [hunk 4]
```diff
@@ -911,6 +917,9 @@ GdipSetLineWrapMode (GpLineGradient *brush, GpWrapMode wrapMode)
 	if (!brush || (wrapMode == WrapModeClamp))
 		return InvalidParameter;
 
+	if (wrapMode < WrapModeTile || wrapMode > WrapModeClamp)
```
Comment (hughbe) [inline src/lineargradientbrush.c]: GDI+ does not modify the wrap mode if the parameter is invalid

File: tests/testlineargradientbrush.c [hunk 5]
```diff
@@ -0,0 +1,1313 @@
...
+    REAL largeBlend[3] = { 1, 2, 3 };
+    REAL largePositions[3] = { 0, 0.5, 1 };
+
+    GdipCreateLineBrushFromRect (&rect, 10, 11, LinearGradientModeHorizontal, WrapModeTile, &brush);
+
+    status = GdipGetLineBlend (brush, blend, positions, 1);
+    assert (status == Ok);
+    assert (floatsEqual (blend[0], 1));
+    // FIXME: it appears that Windows 10+ versions of GDI+ don't copy anything to positions.
+    // assert (floatsEqual (positions[0], 0));
```
Comment (hughbe) [inline tests/testlineargradientbrush.c]: /cc @mellinoe @akoeplinger 

This passes online with rextester.com/l/cpp_online_compiler_visual (an older version of Windows), but fails on Windows 10.

This is a regression in GDI+

File: tests/testlineargradientbrush.c [hunk 6]
```diff
@@ -0,0 +1,1313 @@
...
+    status = GdipSetLineBlend (brush, lineBlend, linePositions, 3);
+    assert (status == Ok);
+
+    status = GdipSetLinePresetBlend (brush, destBlend2, destPositions2, 2);
+    assert (status == Ok);
+
+    status = GdipGetLineBlendCount (brush, &blendCount);
+    assert (status == Ok);
+    // FIXME: this returns 2 with GDI+.
+    // assert (blendCount == 2);
```
Comment (hughbe) [inline tests/testlineargradientbrush.c]: I believe this is a bug in libgdiplus but am unsure - @akoeplinger what do you think!

**Comment by akoeplinger** [comment]
@hughbe this breaks a few System.Drawing tests: 

```

Tests run: 1855, Passed: 1843, Errors: 0, Failures: 2, Inconclusive: 0
  Not run: 10, Invalid: 0, Ignored: 10, Skipped: 0
Elapsed time: 00:00:02.8190000
Errors and Failures:
1) ZeroHeightRectangleF (MonoTests.System.Drawing.Drawing2D.LinearGradientBrushTest.ZeroHeightRectangleF)
     Expected: <System.ArgumentException>
  But was:  <System.OutOfMemoryException> (Not enough memory to complete operation [GDI+ status: OutOfMemory])
  at System.Drawing.GDIPlus.CheckStatus (System.Drawing.Status status) [0x000b7] in /home/travis/build/mono/mono/mcs/class/System.Drawing/System.Drawing/gdipFunctions.cs:224 
  at System.Drawing.Drawing2D.LinearGradientBrush..ctor (System.Drawing.RectangleF rect, System.Drawing.Color color1, System.Drawing.Color color2, System.Drawing.Drawing2D.LinearGradientMode linearGradientMode) [0x00020] in /home/travis/build/mono/mono/mcs/class/System.Drawing/System.Drawing.Drawing2D/LinearGradientBrush.cs:86 
  at (wrapper remoting-invoke-with-check) System.Drawing.Drawing2D.LinearGradientBrush..ctor(System.Drawing.RectangleF,System.Drawing.Color,System.Drawing.Color,System.Drawing.Drawing2D.LinearGradientMode)
  at MonoTests.System.Drawing.Drawing2D.LinearGradientBrushTest+<>c__DisplayClass73_0.<ZeroHeightRectangleF>b__0 () [0x00000] in /home/travis/build/mono/mono/mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/LinearGradientBrushTest.cs:886 
  at NUnit.Framework.Assert.Throws (NUnit.Framework.Constraints.IResolveConstraint expression, NUnit.Framework.TestDelegate code, System.String message, System.Object[] args) [0x0002f] in /home/travis/build/mono/mono/external/nunit-lite/NUnitLite-1.0.0/src/framework/Assert.cs:626 
  at MonoTests.System.Drawing.Drawing2D.LinearGradientBrushTest.ZeroHeightRectangleF () [0x00049] in /home/travis/build/mono/mono/mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/LinearGradientBrushTest.cs:886 
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in /home/travis/build/mono/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305 
2) ZeroWidthRectangle (MonoTests.System.Drawing.Drawing2D.LinearGradientBrushTest.ZeroWidthRectangle)
     Expected: <System.ArgumentException>
  But was:  <System.OutOfMemoryException> (Not enough memory to complete operation [GDI+ status: OutOfMemory])
  at System.Drawing.GDIPlus.CheckStatus (System.Drawing.Status status) [0x000b7] in /home/travis/build/mono/mono/mcs/class/System.Drawing/System.Drawing/gdipFunctions.cs:224 
  at System.Drawing.Drawing2D.LinearGradientBrush..ctor (System.Drawing.Rectangle rect, System.Drawing.Color color1, System.Drawing.Color color2, System.Drawing.Drawing2D.LinearGradientMode linearGradientMode) [0x00020] in /home/travis/build/mono/mono/mcs/class/System.Drawing/System.Drawing.Drawing2D/LinearGradientBrush.cs:72 
  at (wrapper remoting-invoke-with-check) System.Drawing.Drawing2D.LinearGradientBrush..ctor(System.Drawing.Rectangle,System.Drawing.Color,System.Drawing.Color,System.Drawing.Drawing2D.LinearGradientMode)
  at MonoTests.System.Drawing.Drawing2D.LinearGradientBrushTest+<>c__DisplayClass72_0.<ZeroWidthRectangle>b__0 () [0x00000] in /home/travis/build/mono/mono/mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/LinearGradientBrushTest.cs:878 
  at NUnit.Framework.Assert.Throws (NUnit.Framework.Constraints.IResolveConstraint expression, NUnit.Framework.TestDelegate code, System.String message, System.Object[] args) [0x0002f] in /home/travis/build/mono/mono/external/nunit-lite/NUnitLite-1.0.0/src/framework/Assert.cs:626 
  at MonoTests.System.Drawing.Drawing2D.LinearGradientBrushTest.ZeroWidthRectangle () [0x0002e] in /home/travis/build/mono/mono/mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/LinearGradientBrushTest.cs:878 
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in /home/travis/build/mono/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305 
Results saved as /home/travis/build/mono/mono/mcs/class/System.Drawing/TestResult-net_4_x.xml.
make[1]: *** [run-test-lib] Error 1
make[1]: Leaving directory `/home/travis/build/mono/mono/mcs/class/System.Drawing'
make: *** [do-run-test] Error 1
make: Leaving directory `/home/travis/build/mono/mono/mcs/class/System.Drawing'

```

**Comment by hughbe** [comment]
The native code throws an OOM but the c# code explicitly should throw an AE: https://github.com/dotnet/corefx/blob/master/src/System.Drawing.Common/src/System/Drawing/Drawing2D/LinearGradientBrush.cs#L51-L54

Is mono using LinearGradientBrush from corefx?

**Comment by akoeplinger** [comment]
@hughbe not yet, no: https://github.com/mono/mono/blob/d0c98fab566869e53f008af3db49cac11d92db4c/mcs/class/System.Drawing/System.Drawing.Drawing2D/LinearGradientBrush.cs#L82-L90

**Comment by hughbe** [comment]
OK. Is this PR blocked until then?

**Comment by akoeplinger** [comment]
yeah I can't merge the PR if it breaks the tests :)

Can we do the following in the meantime:

```c
if (rect->Width == 0.0 || rect->Height == 0.0)
	return InvalidParameter; // FIXME: should be OutOfMemory to match GDI+ but needs updates in Mono System.Drawing
```

File: tests/testlineargradientbrush.c (hunk 5 above)
Comment (akoeplinger) [reply tests/testlineargradientbrush.c]: Hmm, I have no idea where to file a Windows bug, @mellinoe ? 😄

File: tests/testlineargradientbrush.c (hunk 6 above)
Comment (akoeplinger) [reply tests/testlineargradientbrush.c]: it sounds like a bug in libgdiplus to me as well.

**Pull request merged**